hazelcast-jet-hadoop/src/main/java/com/hazelcast/jet/hadoop/impl/ReadHdfsNewApiP.java

```diff
@@ -0,0 +1,293 @@
+            @Nonnull InputFormat inputFormat,
+            @Nonnull List<InputSplit> splits,
+            @Nonnull BiFunctionEx<K, V, R> projectionFn
+    ) {
+        this.configuration = configuration;
+        this.inputFormat = inputFormat;
+        this.trav = traverseIterable(splits)
+                .flatMap(this::traverseSplit);
+        this.projectionFn = projectionFn;
+    }
+
+    @Override
+    protected void init(@Nonnull Context context) {
+        HazelcastInstanceImpl instance = (HazelcastInstanceImpl) context.jetInstance().getHazelcastInstance();
+        serializationService = instance.getSerializationService();
+    }
+
+    @Override
+    public boolean isCooperative() {
+        return false;
+    }
+
+    @Override
+    public boolean complete() {
+        return emitFromTraverser(trav);
+    }
+
+    @SuppressWarnings("unchecked")
+    private Traverser<R> traverseSplit(InputSplit split) {
+        RecordReader<K, V> reader;
+        try {
+            TaskAttemptContextImpl attemptContext = new TaskAttemptContextImpl(configuration, new TaskAttemptID());
+            reader = inputFormat.createRecordReader(split, attemptContext);
+            reader.initialize(split, attemptContext);
+        } catch (IOException | InterruptedException e) {
+            throw sneakyThrow(e);
+        }
+
+        return () -> {
+            try {
+                while (reader.nextKeyValue()) {
+                    // we clone the key/value if configured so because some of the
+                    // record-readers return the same object for `reader.getCurrentKey()`
+                    // and `reader.getCurrentValue()` which is mutated for each `reader.nextKeyValue()`
+                    K key = copyIfConfigured(reader.getCurrentKey());
+                    V value = copyIfConfigured(reader.getCurrentValue());
+                    R projectedRecord = projectionFn.apply(key, value);
+                    if (projectedRecord != null) {
+                        return projectedRecord;
+                    }
+                }
+                reader.close();
+                return null;
+            } catch (Exception e) {
+                throw sneakyThrow(e);
+            }
+        };
+    }
+
+    @SuppressWarnings("unchecked")
+    private <T> T copyIfConfigured(T o) throws IOException, IllegalAccessException, InstantiationException {
+        if (!configuration.getBoolean(COPY_ON_READ, true)) {
+            return o;
+        }
+        if (o instanceof Writable) {
+            ((Writable) o).write(cloneBufferDataOutput);
+            Writable newO = (Writable) o.getClass().newInstance();
+            newO.readFields(cloneBuffer.getDataInputStream());
+            o = (T) newO;
+            cloneBuffer.reset();
+            return o;
+        }
+        return serializationService.toObject(serializationService.toData(o));
+    }
+
+    private static InputFormat getInputFormat(Configuration configuration) {
+        try {
```

Comment:
you can remove this try block and replace the call to getInputFormat() with `uncheckCall()`